### ISSUE TITLE
Ignore GET response with unexpected observe option

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#688]
There is a behavior change from californium 2.0.0-M9 to 2.0.0-M10.

If we receive a Separated response with an observe option from a GET (without observe option) we will ignore it. (see [UdpMatcher](https://github.com/eclipse/californium/blob/2.0.x/californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java#L290))
Just to be clear, The Token match but there is just an unexpected observe option.

I think that was not the case before. The changes seems to be done in 25548e29afb011e4bf1257b0bb856a985b8a9179.

I totally agree that the device should not answer with the observe option but, 
1) should we be so strict ? I mean if there is other option than observe that we don't expect we will not ignore the response ?
2) Maybe there is a technical issue which explain we change the behavior like this ?

I tried to search in the spec and didn't find anything about that.

To be honest, we currently have some "buggy" devices which exactly behave like this and so which does not work anymore with californium 2.0.0-M10. We currently revert to 2.0.0-M9 waiting to find a solution (or workaround) for this.

**Comment by boaks** [comment]
> If we receive a Separated response with an observe option from a GET (without observe option) we will ignore it. 

>I think that was not the case before. The changes seems to be done in 25548e2.

```
if (response.isNotification() && response.getType() != Type.ACK && !exchange.getRequest().isObserve()) {
	// notification for none observe request
	// including observation cancel request
	LOGGER.debug("ignoring notify {}!", response);
	return null;
}
```

Yes, analysing test with "proactive" cancel observes (GET Observe 1) shows, that a notify should not complete that exchange, otherwise it will not be ensured, that the cancel is really executed.

>I totally agree that the device should not answer with the observe option but,
  
I hope, this is just a notification and the real response is on the way. Or did the device just add the option even to a GET with a new token? If it just adds the option accidentally, does it send an ACK before? Maybe we change it again, that it only ignores it, if the exchange didn't receive ACK before.

**Comment by boaks** [comment]
Or we check, if the token is contained in the `ObservationStore`. Only if it's in there, we ignore it in that case.
So the questions are:
is  there an observation with the same token?
is a ACK send (usually) ahead?

**Comment by sbernard31** [comment]
Unfortunately, the device is pretty stupid..
You send it a GET request without observe option.
It send you a ACK with the right MID.
Then a seperated CON response with the right token but with an unexpected observe option ...
(We never send it GET with observe option before...)

**Comment by boaks** [comment]
OK, then I think, it could be changed to not ignore it, when the request is already ACKed.

**Comment by sbernard31** [comment]
Could you tell me which test about  "proactive" cancel you are talking about ?

**Comment by sbernard31** [comment]
I asked you that because I was hoping that was a unit test and so I could better understand the issue by reading it.
Unfortunately this is not the case, so could you explain me the issue about procancel because I can not see how checking if the request is acknowledged could solve it :-/

**Comment by boaks** [comment]
The point is, that based on the same token, the server may send notifies, which "cross" the request to cancel the observe (with the same token).

```
---- GET CON T1, OBS=0  --->
<------ ACK, T1, OBS 11 ------
<------ NON, T1, OBS 24 -----
<------ NON, T1, OBS 42 -----
<------ NON, T1, OBS 43 -----
---- GET CON T1, OBS=1  
<------ NON, T1, OBS 53 -----
       GET CON T1, OBS=1  --->
<-------- ACK, T1, ------------ 
```

So, if the NON completes the exchange, it's not ensured, that GET CON T1, OBS=1 reaches the server.

**Comment by sbernard31** [comment]
Ok, so why not just making the absence of observe option only mandatory for cancel request (that mean request with observe option = 1) ?

The is more or less what the specification say : _"The resulting response MUST NOT include an Observe Option"_ ?

**Comment by boaks** [comment]
OK. I'll change it on Monday.

**Comment by sbernard31** [comment]
I tried to create a unit test corresponding to the scenario you described. (https://github.com/eclipse/californium/commit/20cd645617df789702420b9c67d67118a87027eb)

```
CON [MID=64974, T=[31f9f4de2e609d18]], GET, /test, observe(0)    ----->
<-----   ACK [MID=64974, T=[31f9f4de2e609d18]], 2.05, observe(1)
(First notification)
<-----   CON [MID=8001, T=[31f9f4de2e609d18]], 2.05, observe(3)
ACK [MID=8001]   ----->
(Send pro active cancel)
CON [MID=64975, T=[31f9f4de2e609d18]], GET, /test, observe(1)    ----->
(Second notification)
<-----   CON [MID=8002, T=[31f9f4de2e609d18]], 2.05, observe(4)
RST [MID=8002]   -----> ( Surprising ?!)
(Third notification)
<-----   NON [MID=8003, T=[31f9f4de2e609d18]], 2.05, observe(5)
RST [MID=8003]   -----> ( Surprising ?!)
(Response of pro active cancel, ACK is lost)
<-----   CON [MID=64975, T=[31f9f4de2e609d18]], 2.05
RST [MID=64975]   -----> ( Surprising ?!)
```

Strangely, when we send a proactive cancel request, next notifications will receive a RST response even if I never cancel the observation.

I think this is a side effect as cancel request and observation use the same token (as required by the spec)
Proactive cancel exchange is store in memory exchange store and  Observation store in observationStore.
As we first search in memory exchange store, once there is an proactive cancel exchange we never search in observation store.

The last RST is even more suprising.

**Comment by boaks** [comment]
> Strangely, when we send a proactive cancel request, next notifications will receive a RST response even if I never cancel the observation.

I'm not sure, why a RST or a "not processed" CON should be strange.

**Comment by sbernard31** [comment]
(The last RST was an error in my tests)
Ok I'll try to explain better. I add a notification after the proactive cancel to better expose the "potential issue"

At client side, when we establishes an observe relation,  we add a new observation in observation store.

To cancel it, we must use `endpoint.cancelObservation(observeRequestToken)`. This removes observation. After that all notifications will be answered with a RST.

In my case, I do not "cancel" the observation, I just send a proactive cancel request which should remove the relation at server side but not at client side, so I expected that notification was still handled. (I do no use coapclient, just endpoint API)

Here is what I expected : 
```
------- GET CON M1, T1, OBS=0 ---> // observe request
<------ ACK, M1, T1, OBS 11 ------
<------ CON, M2, T1, OBS 24 ------ // 1st notification
------- ACK M2              ----->
------- GET CON M3, T1, OBS=1 ---> // send proactive cancel
<------ CON M4, T1, OBS 42  ------ // 2nd notification
------- ACK M4              ----->
<------ NON M5, T1, OBS 43  ------ // 3rd notification
X------ ACK M3              ------ // proactive ACK lost
<------ CON M6, T1          ------ // response of proactive cancel
------- ACK M6              ---- >
<------ CON M7, T1, OBS 42  ------ // 4th notification
------- ACK M7              ----->
```
and I get : 
```
------- GET CON M1, T1, OBS=0 ---> // observe request
<------ ACK, M1, T1, OBS 11 ------
<------ CON, M2, T1, OBS 24 ------ // 1st notification
------- ACK M2              ----->
------- GET CON M3, T1, OBS=1 ---> // send proactive cancel
<------ CON M4, T1, OBS 42  ------ // 2nd notification
------- RST M4              ----->
<------ NON M5, T1, OBS 43  ------ // 3rd notification
------- RST M5              ----->
X------ ACK M3              ------ // proactive ACK lost
<------ CON M6, T1          ------ // response of proactive cancel
------- RST M6              ----->
<------ CON M7, T1, OBS 42  ------ // 4th notification
------- ACK M7              ----->

```

First time, I see this result I was a bit surprise but like you I thought "ok why not, we don't handle notification as we cancel it by sending a proactive cancel request", then I look at the code and I believe the is just a side effect because exchange/observe/observation store refactoring.

Proactive cancel exchange is store in InMemoryExchangeStore and observation store in ObservationStore.
When we receive a notification, we first search in InMemoryExchangeStore then in ObservationStore. Once there is an ongoing proactive cancel exchange we don't search in observation store so all notification are ignored until this exchange was completed.

Maybe this side effect is acceptable as generally when you send a proactive cancel request you will cancel observation before. But as I think this behavior was not intended, I just raise the issue here.

**Comment by boaks** [comment]
It's a side effect of "keep the cancel observe exchange" open. If we forward the notify, the current `StackTopAdapter.receiveResponse()` would complete the exchange, and so it would be required to sort out the special case there. To keep it simple, I assumed, that after cancel, there is no interest in notifies and so it's much easier to stop them in the matcher. 

> DEBUG org.eclipse.californium.core.network.UdpMatcher - ignoring notify CON-2.05   MID= 8002, Token=[cfeb4f36e6939558], OptionSet={"Observe":4}, "556832279332579561434386".. 32 bytes!
> 
> DEBUG org.eclipse.californium.core.network.UdpMatcher - ignoring notify NON-2.05   MID= 8003, Token=[cfeb4f36e6939558], OptionSet={"Observe":5}, "556832279332579561434386".. 32 bytes!

Do you have any issue or concern with stopping these notifies while the cancel exchange is pending? 
PR #689 is already limiting this "notify ignore" to cancel observe request, so I feel better with stop them in the matcher.

**Comment by sbernard31** [comment]
> To keep it simple, I assumed, that after cancel, there is no interest in notifies and so it's much easier to stop them in the matcher.

I understand. This is just a bit strange that after proactive cancel exchange is completed,  Notify is handleable again.
But I agree this is not a big deal and fixing this could bring more complexity.

**Comment by boaks** [comment]
> This is just a bit strange that after proactive cancel exchange is completed, Notify is handleable again.

It's completed with a "retransmission timeout" without response. You don't know, if the cancel request made it to the server. And in your use-case, your server seems refuse to stop notifies by RST and by cancel request. It still continues to send notifications. I guess the best approach for a client would be just resend the cancel again. 
But with your coap-server, obviously ignoring RST for notifies and cancel request, I guess even that will be in vain :-).

**Comment by sbernard31** [comment]
I think we can close this issue now. (fixed in **2.0.x** by f4af90092650184421f1c86b166a4060029bb796 tested by f605ca043d1bcafe0f8809a9e7baf8619054da79)
